Add scale & PDF information to the DQM histograms

SanghyunKo opened the pull request:
Updated generator DQM histogram code to include scale & PDF information.

Currently, generator DQM provides little information on scale & PDFs, and the only variable included in DQM histogram from HepMC::PdfInfo is Bjorken x. This update makes DQM histograms to provide scale & PDF information so that validators can check this information for release validation & generator validation. Relevant talk in PH Generator meeting to this proposal can be found at [1].

Tested on CMSSW_10_2_X_2018-06-15-1100.

[1] [https://indico.cern.ch/event/725198/contributions/3026963/attachments/1661134/2661397/180604_shKo.pdf](https://indico.cern.ch/event/725198/contributions/3026963/attachments/1661134/2661397/180604_shKo.pdf)

SanghyunKo committed:
Add scale & PDF information to the DQM histograms

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23601/5210

cmsbuild commented:
A new Pull Request was created by  @SanghyunKo (Sanghyun Ko) for master.

It involves the following packages:

Validation/EventGenerator


@efeyazgan, @vazzolini, @kmaeshima, @perrozzi, @dmitrijus, @cmsbuild, @jfernan2, @alberto-sanchez, @vanbesien can you please review it and eventually sign? Thanks.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

alberto-sanchez commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 980c20b4bea0a6b1c39cb184f333474fa36ca8af
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28718/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28718/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902004
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901811
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 95.875 KiB( 30 files compared)
 - DQMHistoSizes: changed ( 1306.0,... ): 7.375 KiB        Generator/Particles
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

alberto-sanchez commented:
+1

dmitrijus commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented on Validation/EventGenerator/plugins/BasicHepMCValidation.cc:
```diff
@@ -173,12 +184,41 @@ void BasicHepMCValidation::analyze(const edm::Event& iEvent,const edm::EventSetu
   if(pdf){
     bjorken = ((pdf->x1())/((pdf->x1())+(pdf->x2())));
+    logQScale = log10(pdf->scalePDF());
+    if(logQScale > 4.) logQScale = 3.995; // visualize overflow & underflow in the histograms
```
for maintainability would not be better to have a parameter that sets both the limits of the histogram range and the safety value here?

SanghyunKo replied:
Okay. I'll improve that part by introducing some global parameters for upper bound & lower bound of the histograms.

SanghyunKo committed:
Improved maintainability

Improved maintainability by introducing global parameters for upper bound and lower bound of the PDF & scale related histograms.

Tested on CMSSW_10_2_X_2018-06-19-2300.

SanghyunKo committed:
Improved maintainability

Improved maintainability by introducing global parameters for upper bound and lower bound of the PDF & scale related histograms.

Tested on CMSSW_10_2_X_2018-06-19-2300.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23601/5267

cmsbuild commented:
Pull request #23601 was updated. @efeyazgan, @vazzolini, @kmaeshima, @perrozzi, @dmitrijus, @cmsbuild, @jfernan2, @alberto-sanchez, @vanbesien can you please check and sign again.

alberto-sanchez commented:
please test

cmsbuild commented:
+1
Tested at: 8264c589933853b0c0494d02b8b31321df99cb7b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28783/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8264c589933853b0c0494d02b8b31321df99cb7b
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28783/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28783/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23601/28783/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902133
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901940
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 95.875 KiB( 30 files compared)
 - DQMHistoSizes: changed ( 135.4,... ): 7.375 KiB        Generator/Particles
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

alberto-sanchez commented:
+1

fabiocos commented:
+1

dqm had already signed it

fabiocos commented:
merge

andrius-k commented:
+1

Signing for the sake of completeness.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
